Sources/GRPCCore/Metadata.swift

```diff
@@ -14,5 +14,329 @@
  * limitations under the License.
  */
 
-// FIXME: placeholder.
-public typealias Metadata = [String: String]
+/// A collection of metadata key-value pairs, found in RPC streams.
+/// A key can have multiple values associated to it.
+/// Values can be either strings or binary data, in the form of `Data`.
+/// - Note: Binary values must have keys ending in `-bin`, and this will be checked when adding pairs.
+public struct Metadata: Sendable, Hashable {
+
+  /// A metadata value. It can either be a simple string, or binary data.
+  public enum Value: Sendable, Hashable {
+    case string(String)
+    case binary([UInt8])
+  }
+
+  /// A metadata key-value pair.
+  internal struct KeyValuePair: Sendable, Hashable {
+    internal let key: String
+    internal let value: Value
+
+    /// Constructor for a metadata key-value pair.
+    /// - Parameters:
+    ///   - key: The key for the key-value pair.
+    ///   - value: The value to be associated to the given key. If it's a binary value, then the associated
+    ///   key must end in `-bin`, otherwise, this method will produce an assertion failure.
+    init(key: String, value: Value) {
+      if case .binary = value {
+        assert(key.hasSuffix("-bin"), "Keys for binary values must end in -bin")
+      }
+      self.key = key
+      self.value = value
+    }
+  }
+
+  public typealias KeyValue = (key: String, value: Value)
+
+  private var elements: [KeyValuePair]
+
+  /// The Metadata collection's capacity.
+  public var capacity: Int {
+    self.elements.capacity
+  }
+
+  /// Initialize an empty Metadata collection.
+  public init() {
+    self.elements = []
+  }
+
+  /// Reserve the specified minimum capacity in the collection.
+  /// - Parameter minimumCapacity: The minimum capacity to reserve in the collection.
+  public mutating func reserveCapacity(_ minimumCapacity: Int) {
+    self.elements.reserveCapacity(minimumCapacity)
+  }
+
+  /// Add a new key-value pair, where the value is a string.
+  /// - Parameters:
+  ///   - stringValue: The string value to be associated with the given key.
+  ///   - key: The key to be associated with the given value.
+  public mutating func addString(_ stringValue: String, forKey key: String) {
+    self.addValue(.string(stringValue), forKey: key)
+  }
+
+  /// Add a new key-value pair, where the value is binary data, in the form of `Data`.
+  /// - Parameters:
+  ///   - binaryValue: The `Data` to be associated with the given key.
+  ///   - key: The key to be associated with the given value. Must end in `-bin`.
```

Comment:
Super nit: we're talking about a string so should be double quoted rather than backticks:

```suggestion
  ///   - key: The key to be associated with the given value. Must end in "-bin".
```